### PR TITLE
3868 GAPS -> asset properties

### LINKED ISSUE #3868: GAPS: Asset Properties
Currently we only have the ability to define custom properties against the the asset catalogue.
However, with the GAPS work, we need to be able to associate additional data with the asset itself, primarily when 

This requires a database table `Asset Property`, Which defines what properties can be assigned to an asset.
The a properties themselves are stored in a `properties` field as JSON.

## Acceptance Criteria
- [x] Graphql for viewing available `asset_property` records is available
- [x] asset_properties are available via the asset graphql object
- [x] The system shouldn't allow duplicate records for an asset item property, e.g there should only be 1 `donor` record for each asset
- [x] `asset_properties` added centrally should be synced to all sites (Central data)
- [x] Any new `properties` added to an asset at a remote site should be synced to OMS Central

## DB Design

```mermaid
erDiagram
    asset {
        TEXT id PK "NOT NULL"
        TEXT properties "JSON STRING"
    }
    asset_property {
        TEXT id PK "NOT NULL"
        TEXT name "NOT NULL (UNIQUE)"
        TEXT asset_class_id "NULL = any class"
        TEXT asset_category_id "NULL = any category"
        TEXT asset_type_id "NULL = any type"
        TEXT value_type "ENUM 'STRING','BOOLEAN','INTEGER','FLOAT'"
        TEXT allowed_values 
    } 
```

---

| User story  | A.1.2 |
| :------------- | :------------- |
As a… | Health Care Worker (HCW) or warehouse worker
I want… | To be able to easily add my CCE to the system
So that… | I don’t need to type in a lot of data manually when registering my facility’s CCE 

**Solution A.1.2**
The system will allow users to register their CCE using a barcode scanner. 

The scanner will work with UNICEF’s proposed 2D DataMatrix format (see ‘Perf spec E003 BC01.1 Global asset identification - Proposed revision Nov 23’ PDF) by detecting the PQS product number in the 2D DataMatrix and importing the remaining data from the corresponding product entry in the WHO PQS Catalogue. 

The serial number will also be imported from the 2D DataMatrix. 

Further support for the 2D DataMatrix may be added as necessary once the proposed standard has been revised, detailed and approved by WHO PQS.

Users will also be able to easily search for and select my make/model of CCE from a list in the event that they can not / do not want to use the barcode scanning functionality. 

The following data will be captured for each WHO PQS CCE:

- E001 - Cold rooms and freezer rooms
  - Supplied by current WHO PQS Catalogue 
    - PQS code
    - Type
    - Class
    - Category
    - Manufacturer
    - Model
    - Energy source
    - Refrigerant type(s)
  - Not supplied by current WHO PQS Catalogue
    - Can be input manually by user
      - Asset number (country’s unique asset identifier)
      - Installation date
    - **Can be retrieved from 2D DataMatrix or input manually by user**
      - **Serial number**
      - **External dimensions - WxDxH (in cm)**
      - **Storage capacity +5 °C (in litres)**
      - **Storage capacity -20 °C (in litres)**
      - **Storage capacity -70 °C (in litres)**
      - **Waterpack storage capacity (in Kg)**
      - **Waterpack freezing capacity per 24 hours (in Kg)**
      - **Energy consumption (stable running, continuous power) (in KW per day)**
      - **Energy consumption during freezing (in KW per day)**
      - **Hold over time (hours)**
      - **Climate zone**
      - **Freeze protection** 
      - **Temperature monitoring device (integrated, external or none)**
      - **Voltage stabilizer (integrated, external or none)**
      - **Warranty period (start-end date)** 
      - **Links to:** 
        - **PQS Specification sheet**
        - **Manufacturer’s Operating Manual**
        - **Manufacturer’s Warranty**
        - **Manufacturer’s Maintenance Instructions**

---
| User story  | A.1.2.2 |
| :------------- | :------------- |
As a… | Health Care Worker (HCW) or warehouse worker
I want… | To be able to specify the volume for each temperature range for any WHO PQS CCE that does not already include this information by default when I add it to my registry - e.g. E001 cold rooms that are custom built
So that… | The volumes are available for use in cold storage capacity and demand calculations when the CCE is added by HCW to their CCE registry.  

**Solution A.1.2.2**
The system will allow users to specify a volume when registering CCE in the WHO PQS Category E001 - ‘Cold rooms and freezer rooms’.

---

Moneyworks Jobcode 🧰
CCEIDEV - development

### PR DESCRIPTION
**Pull request by jmbrunskill**
Fixes #3868

# 👩🏻‍💻 What does this PR do?

Adds properties that can be assigned at an asset level rather than asset catalogue item.
If the approach is agreed, I'll refactor the asset catalogue item to use the same pattern.

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] This will be tested as part of the epic

# 📃 Documentation

- [X] **Part of an epic**: documentation will be completed for the feature as a whole

### CODE REVIEW

File: server/graphql/asset/src/lib.rs [hunk 1]
```diff
@@ -97,76 +92,6 @@ impl AssetMutations {
     }
 }
 
-#[derive(Default, Clone)]
```
Comment (jmbrunskill) [inline server/graphql/asset/src/lib.rs:100]: Moved all these to `asset_log` as the file was getting unwieldy

File: client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx [hunk 2]
```diff
@@ -246,12 +246,15 @@ export const Summary = ({ draft, onChange, locations }: SummaryProps) => {
...
         </Section>
-        {draft.properties.length === 0 ? null : (
+        {draft.parsedProperties ? null : (
           <Section heading={t('label.catalogue-properties')}>
-            {draft.properties.map(property => (
-              <Row key={property.id} label={property.name}>
+            {Object.keys(draft.parsedProperties).map(key => (
+              <Row
+                key={key}
+                label={draft.parsedProperties[key]?.toString() ?? ''}
```
Comment (jmbrunskill) [inline client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx:254]: Is this ok? I figure it probalby is?

File: client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx [hunk 3]
```diff
@@ -73,11 +74,17 @@ export const EquipmentDetailView = () => {
 
   useEffect(() => {
     if (!data) return;
+    let properties = JSON.parse(data.properties);
```
Comment (jmbrunskill) [inline client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx:77]: Should we merge the asset catalogue properties with the asset properties here? Or should we do it in the graphql layer?

File: client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx [hunk 4]
```diff
@@ -14,7 +14,7 @@ import {
   useIsCentralServerApi,
 } from '@openmsupply-client/common';
 import { Status } from '../../Components';
-import { formatPropertyValue } from '../../utils';
+// import { formatPropertyValue } from '../../utils';
```
Comment (jmbrunskill) [inline client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx:17]: I guess we get rid of this eventually...

File: server/graphql/asset/src/types/asset.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    //                 .collect()
+    //         }
+    //         None => vec![],
+    //     };
+
+    //     Ok(properties)
+    // }
+
+    pub async fn properties(&self) -> Result<String> {
+        // TODO: Merge properties with catalogue item properties
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset.rs:224]: Obviously this looses all the asset catalogue item properties.
Was going to merge here, but wonder if we should return both as different fields and merge in the frontend.
That would also let us show if something's been overridden for example?

File: server/graphql/asset/src/types/asset_catalogue_item.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
+use async_graphql::*;
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset_catalogue_item.rs:1]: This is just moved to re-arrange.

File: server/graphql/asset/src/types/asset_log.rs [hunk 7]
```diff
@@ -0,0 +1,356 @@
+use std::vec;
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset_log.rs:1]: This is just moved to re-arrange

File: server/graphql/asset/src/logs/mod.rs [hunk 8]
```diff
@@ -8,3 +8,82 @@ mod query_reason;
 pub use query_reason::*;
 mod delete_reason;
 pub use delete_reason::*;
+
```
Comment (jmbrunskill) [inline server/graphql/asset/src/logs/mod.rs:11]: This is just moved to re-arrange

File: server/service/src/sync/test/test_data/asset.rs [hunk 9]
```diff
@@ -34,6 +34,7 @@ fn asset1() -> AssetRow {
         created_datetime: Defaults::naive_date_time(),
         modified_datetime: Defaults::naive_date_time(),
         deleted_datetime: None,
+        properties: None,
```
Comment (jmbrunskill) [inline server/service/src/sync/test/test_data/asset.rs:37]: Should create another record with the embedded json to test here...

File: server/graphql/asset/src/types/asset.rs [hunk 10]
```diff
@@ -0,0 +1,343 @@
+use std::vec;
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset.rs:1]: This has mostly been moved too

File: server/graphql/asset/src/types/asset.rs [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+        let asset_id = &self.row().id;
+        let loader = ctx.get_loader::<DataLoader<SyncFileReferenceLoader>>();
+        let result_option = loader.load_one(asset_id.to_string()).await?;
+
+        let documents = SyncFileReferenceConnector::from_vec(result_option.unwrap_or(vec![]));
+
+        Ok(documents)
+    }
+
+    pub async fn catalog_properties(
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset.rs:204]: renamed

File: server/graphql/asset/src/types/asset.rs [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+                    .into_iter()
+                    .collect()
+            }
+            None => vec![],
+        };
+
+        Ok(properties)
+    }
+
+    pub async fn properties(&self) -> Result<String> {
```
Comment (jmbrunskill) [inline server/graphql/asset/src/types/asset.rs:226]: Now asset properties only.

File: server/repository/src/db_diesel/assets/types.rs [hunk 13]
```diff
@@ -0,0 +1,12 @@
+use diesel_derive_enum::DbEnum;
+use serde::{Deserialize, Serialize};
+
+#[derive(DbEnum, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+#[DbValueStyle = "SCREAMING_SNAKE_CASE"]
+pub enum PropertyValueType {
```
Comment (jmbrunskill) [inline server/repository/src/db_diesel/assets/types.rs:7]: Shared type now, saves extra types in Rust and in Postgres

File: client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx (hunk 2 above)
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx:268]: Will be updated in Next PR anyway...

File: client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx [hunk 14]
```diff
@@ -73,11 +74,17 @@ export const EquipmentDetailView = () => {
...
+    if (!properties || !isObject(properties)) {
+      // Invalid properties, reset to empty object
+      properties = {};
+    }
     setDraft({
       ...data,
       locationIds: draft?.locationIds
         ? draft.locationIds
         : data.locations.nodes.map(location => location.id),
+      parsedProperties: properties,
```
Comment (mark-prins) [inline client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx:87]: the reason for renaming the prop is to get the different type is it?

File: server/repository/src/migrations/v2_01_00/assets/asset_property.rs [hunk 15]
```diff
@@ -0,0 +1,40 @@
+use crate::{
+    migrations::{sql, JSON},
+    StorageConnection,
+};
+
+pub(crate) fn migrate(connection: &StorageConnection) -> anyhow::Result<()> {
+    const PROPERTY_VALUE_TYPE: &str = if cfg!(feature = "postgres") {
+        "PROPERTY_VALUE_TYPE" // This is created as part of the asset_catalogue_property migration
```
Comment (mark-prins) [inline server/repository/src/migrations/v2_01_00/assets/asset_property.rs:8]: This probably doesn't matter.. but for completeness
```suggestion
        "property_value_type" // This is created as part of the asset_catalogue_property migration
```

**Review by mark-prins — APPROVED** [review]
Nice! that looks good - thanks James

File: client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx [hunk 16]
```diff
@@ -23,6 +23,7 @@ import { StatusLogs } from './Tabs/StatusLogs';
 import { Documents } from './Tabs/Documents';
 import { ActivityLogList, useLocation } from '@openmsupply-client/system';
 import { DraftAsset } from '../types';
+import { isObject } from 'lodash';
```
[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx:26]: Nice! Was wondering if lodash was a bad idea....

File: client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx (hunk 14 above)
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx:87]: Yes, do you have a better idea? Slightly hacky I know...

File: client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx (hunk 2 above)
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx:268]: I feel like parsed communicates the intention better.
When I refactor catalogueProperties I'll revisit...

File: client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx (hunk 14 above)
Comment (mark-prins) [reply client/packages/coldchain/src/Equipment/DetailView/DetailView.tsx:87]: slightly hacky maybe, but I think the best approach. The use of the `draft` object is a bit wild west anyway

File: client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx [hunk 17]
```diff
@@ -259,6 +259,24 @@ export const Summary = ({ draft, onChange, locations }: SummaryProps) => {
             ))}
           </Section>
         )}
+
+        {!draft.parsedProperties ? null : (
+          <Section heading={t('label.asset-properties')}>
+            {Object.keys(draft.parsedProperties).map(key => (
+              <Row
+                key={key}
+                label={draft.parsedProperties[key]?.toString() ?? ''}
```
[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx]: Sorry for being pedantic @mark-prins @jmbrunskill, wanted to trace the reason for compilation failure in develop

[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx]: 🤔 Is there something that needs changing? Sorry, didn't understand the comment

[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx]: ohh just that it lead to stray `>` which caused compilation problem in develop

[the remark this answers is not in the document]
Comment (jmbrunskill) [reply client/packages/coldchain/src/Equipment/DetailView/Tabs/Summary.tsx]: Sorry, thought I'd checked after commiting the suggestion but maybe not! Promptly changed it in my next PR which is probably why I didn't notice!

**Pull request merged**